test/unit/Reflection/ReflectionMethodTest.php

```diff
@@ -810,4 +811,71 @@ public function testWithCurrentClass(): void
+
+        $hookProperty = $getClassInfo->getProperty('hookWithImplicitParameter');
+        self::assertTrue($hookProperty->hasHook(ReflectionPropertyHookType::Set));
+
+        $hookReflection = $hookProperty->getHook(ReflectionPropertyHookType::Set);
+        self::assertInstanceOf(ReflectionType::class, $hookReflection->getReturnType());
+        self::assertSame('void', $hookReflection->getReturnType()->getName());
+
+        self::assertSame(1, $hookReflection->getNumberOfParameters());
+        self::assertInstanceOf(ReflectionParameter::class, $hookReflection->getParameter('value'));
+
+        $hookParameterReflection = $hookReflection->getParameter('value');
+        self::assertSame(0, $hookParameterReflection->getPosition());
+        self::assertFalse($hookParameterReflection->isOptional());
+        self::assertInstanceOf(ReflectionType::class, $hookParameterReflection->getType());
+        self::assertSame('string', $hookParameterReflection->getType()->getName());
+    }
+
+    public function testSetPropertyHookHasExplicitParameter(): void
+    {
+        $reflector    = new DefaultReflector(new SingleFileSourceLocator(__DIR__ . '/../Fixture/PropertyHooks.php', $this->astLocator));
+        $getClassInfo = $reflector->reflectClass('Roave\BetterReflectionTest\Fixture\SetPropertyHooksParameters');
+
+        $hookProperty = $getClassInfo->getProperty('hookWithExplicitParameter');
+        self::assertTrue($hookProperty->hasHook(ReflectionPropertyHookType::Set));
+
+        $hookReflection = $hookProperty->getHook(ReflectionPropertyHookType::Set);
+        self::assertInstanceOf(ReflectionType::class, $hookReflection->getReturnType());
+        self::assertSame('void', $hookReflection->getReturnType()->getName());
+
+        self::assertSame(1, $hookReflection->getNumberOfParameters());
+        self::assertInstanceOf(ReflectionParameter::class, $hookReflection->getParameter('value'));
+
+        $hookParameterReflection = $hookReflection->getParameter('value');
+        self::assertSame(0, $hookParameterReflection->getPosition());
+        self::assertFalse($hookParameterReflection->isOptional());
+        self::assertInstanceOf(ReflectionType::class, $hookParameterReflection->getType());
+        self::assertSame('int', $hookParameterReflection->getType()->getName());
+    }
+
+    /** @return list<array{0: non-empty-string, 1: string|null}> */
+    public static function getPropertyHookReturnTypeProvider(): array
+    {
+        return [
+            ['hookWithString', 'string'],
+            ['hookWithInteger', 'int'],
+            ['hookWithUnion', 'string|int'],
+            ['hookWithoutType', null],
+        ];
+    }
+
+    #[DataProvider('getPropertyHookReturnTypeProvider')]
+    public function testGetPropertyHookReturnType(string $propertyName, string|null $returnType): void
+    {
+        $reflector = new DefaultReflector(new SingleFileSourceLocator(__DIR__ . '/../Fixture/PropertyHooks.php', $this->astLocator));
```

Comment:
One note on the design of these tests: having everything inside a third-party file makes them a bit harder to locate when jumping around source, compared to using a string locator inlined with the test